### PR TITLE
GH-39214: [Java] Support reproducible build

### LINKED ISSUE #146: [Java] Provide reproducible builds
### Describe the enhancement requested

Currently, our provided artifacts are not reproducible, meaning that two builds provide different artifacts.

Reproducible builds, also known as deterministic compilation, is a process of compiling software which ensures the resulting binary code can be reproduced. Source code compiled using deterministic compilation will always output the same binary.

In order to have reproducible builds, we need to add a property (with fixed compile output timestamp) and be sure we use the maven plugins version compliant with reproducible builds.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by jbonofre**
### Rationale for this change

This PR add reproducible builds support.

### Are these changes tested?

I tested the builds are reproducible on `arrow-vector`, `arrow-memory`, ... using `mvn clean verify artifact:compare`.

### Are there any user-facing changes?

No
* Closes: apache/arrow-java#146
* GitHub Issue: #146

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
Thanks!

**Comment by jbonofre** [comment]
@lidavidm let me check why GH Action is not happy 😃

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

File: java/pom.xml [hunk 1]
```diff
@@ -28,6 +28,7 @@
   <url>https://arrow.apache.org/</url>
 
   <properties>
+    <project.build.outputTimestamp>2023-12-13T00:00:00Z</project.build.outputTimestamp>
```
Comment (davisusanibar) [inline java/pom.xml:31]: The parameter 'project.build.outputTimestamp' is defined at the level of the parent project. Should it also be added to child projects?

File: java/format/pom.xml [hunk 2]
```diff
@@ -23,6 +23,10 @@
 <name>Arrow Format</name>
 <description>Generated Java files from the IPC Flatbuffer definitions.</description>
 
+  <properties>
+    <project.build.outputTimestamp>2023-12-13T00:00:00Z</project.build.outputTimestamp>
+  </properties>
```
Comment (davisusanibar) [inline java/format/pom.xml:28]: Can you provide some commands or steps for validating a reproducible build, such as in the Arrow Format module?

Comment (davisusanibar) [reply java/format/pom.xml:28]: When I tried these commands, it finished ok with or without any changes:

````
$ mvn clean install
$ mvn clean verify artifact:compare
...
[INFO] Saved info on build to /Users/dsusanibar/fork/arrow/java/format/target/arrow-format-15.0.0-SNAPSHOT.buildinfo
[INFO] Checking against reference build from central...
[INFO] Reference buildinfo file not found: it will be generated from downloaded reference artifacts
[INFO] Reference build java.version: 11 (from MANIFEST.MF Build-Jdk-Spec)
[INFO] Reference build os.name: Unix (from pom.properties newline)
[INFO] Minimal buildinfo generated from downloaded artifacts: /Users/dsusanibar/fork/arrow/java/format/target/reference/arrow-format-15.0.0-SNAPSHOT.buildinfo
[INFO] Reproducible Build output summary: 5 files ok
[INFO] Reproducible Build output comparison saved to /Users/dsusanibar/fork/arrow/java/format/target/arrow-format-15.0.0-SNAPSHOT.buildcompare
...
````

File: java/pom.xml (hunk 1 above)
Comment (jbonofre) [reply java/pom.xml:31]: It's possible but not recommended: if you do artifact:check-buildplan it will raise an error in the case it's inherited from parent pom.

File: java/format/pom.xml (hunk 2 above)
Comment (jbonofre) [reply java/format/pom.xml:28]: You are right for the check: mvn clean install and then mvn clean verify artifact:compare do a check. 
This PR ensures we have reproducible build, not meaning that some module are not ok already (AFAIR, flight and algorithm were not reproducible).

Comment (jbonofre) [reply java/format/pom.xml:28]: You can also check with artifact:check-buildplan.

File: java/pom.xml (hunk 1 above)
Comment (davisusanibar) [reply java/pom.xml:31]: Thank you:
````
$ mvn clean artifact:check-buildplan 
[ERROR] project.build.outputTimestamp property should not be inherited but defined in POM /Users/dsusanibar/Documents/business/voltron/fork/arrow/java/format/pom.xml
````

If it is required on all projects, it will give us clarity if it is only defined on child projects and not parent project. What do you think about that?

Comment (jbonofre) [reply java/pom.xml:31]: Good point. It makes sense. Let me update the PR accordingly. Thanks !

File: java/format/pom.xml (hunk 2 above)
Comment (lidavidm) [reply java/format/pom.xml:28]: What we might want is to add this to CI somewhere (or possibly the release verification process) so we can make sure it stays working.

File: java/pom.xml (hunk 1 above)
Comment (jbonofre) [reply java/pom.xml]: Done, I include the timestamp property only in modules building jar (not pom).

File: java/format/pom.xml (hunk 2 above)
Comment (jbonofre) [reply java/format/pom.xml:28]: @lidavidm good idea. Let me check to add a workflow in GH actions.

**Comment by jbonofre** [comment]
FYI, I'm resuming work on this PR:
- checking the build failure on CI
- checking to add a GH Action workflow to verify

**Comment by lidavidm** [comment]
> checking to add a GH Action workflow to verify

It might be best to add it to an existing workflow, or again, adding it to the release verification script even (what do you think @assignUser?)

**Comment by jbonofre** [comment]
> > checking to add a GH Action workflow to verify
> 
> It might be best to add it to an existing workflow, or again, adding it to the release verification script even (what do you think @assignUser?)

@lidavidm it makes sense, I do like this. Thanks !

**Comment by assignUser** [comment]
How exactly does this work, it compares the current build against a reference build? How does that work for dev versions?

We do have binary verification scripts so it would make sense to add it there, that would automatically also run it. `dev/release/verify-release-candidate.sh` with `TEST_JARS=1' is run by the jar binary verification script (this downloads previously build nightly jars iirc) cc @raulcd who is a bit more familiar with the release scripts

**Comment by jbonofre** [comment]
@lidavidm yup, it's what I plan to do (and I used while working on this PR 😄 ).

First I will fix the failing checks (AMD64 Conda Integration Test, etc).

**Comment by jbonofre** [comment]
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.4.1:enforce (enforce-maven-version) on project arrow-bom: 
Error:  Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion failed with message:
Error:  Detected Maven Version: 3.5.0 is not in the allowed range [3.6.3,).
Error:  -> [Help 1]
```

That's due to the Apache POM update. I will propose fix as part of this PR.

**Comment by jbonofre** [comment]
I fixed the enforcer range by setting `minimalMavenBuildVersion`.

Let me take a look on the Merge Script workflows.

**Comment by kou** [comment]
When is `<project.build.outputTimestamp>` updated?
Does `mvn versions:set` update it?

It seems that our `dev/release/01-prepare.sh` updates it: https://github.com/apache/arrow/actions/runs/7263564652/job/19789165030?pr=39215#step:7:24478

**Comment by kou** [comment]
https://maven.apache.org/guides/mini/guide-reproducible-builds.html#faq

> if you have a custom release process tooling, you'll need to add the feature to your release tooling. Notice that if you're using versions-maven-plugin in custom release scripts, starting with release 2.9.0, [versions:set goal updates the property](https://github.com/mojohaus/versions-maven-plugin/issues/453).

**Comment by kou** [comment]
It seems that updated value is always the current time: https://github.com/mojohaus/versions/pull/522/files#diff-544db8b7e4906f57cb6c1bd43ba1100760f876da22b46b2ca381ab374cb169faR530

**Comment by kou** [comment]
How about this for "Dev /Source Release and Merge Script" failures?

```diff
diff --git a/dev/release/01-prepare-test.rb b/dev/release/01-prepare-test.rb
index 8fb23f45f0..f4bf01f1f6 100644
--- a/dev/release/01-prepare-test.rb
+++ b/dev/release/01-prepare-test.rb
@@ -264,18 +264,17 @@ class PrepareTest < Test::Unit::TestCase
     end
 
     Dir.glob("java/**/pom.xml") do |path|
-      version = "<version>#{@snapshot_version}</version>"
-      lines = File.readlines(path, chomp: true)
-      target_lines = lines.grep(/#{Regexp.escape(version)}/)
-      hunks = []
-      target_lines.each do |line|
-        new_line = line.gsub(@snapshot_version) do
-          @release_version
+      hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
+        if line.include?("<version>#{@snapshot_version}</version>")
+          new_line = line.gsub(@snapshot_version) do
+            @release_version
+          end
+          [line, new_line]
+        elsif line.include?("<project.build.outputTimestamp>")
+          [line, normalize_pom_xml_output_timestamp(line)]
+        else
+          [nil, nil]
         end
-        hunks << [
-          "-#{line}",
-          "+#{new_line}",
-        ]
       end
       expected_changes << {hunks: hunks, path: path}
     end
diff --git a/dev/release/post-11-bump-versions-test.rb b/dev/release/post-11-bump-versions-test.rb
index 4b6933d610..08a900f71c 100644
--- a/dev/release/post-11-bump-versions-test.rb
+++ b/dev/release/post-11-bump-versions-test.rb
@@ -244,37 +244,19 @@ class PostBumpVersionsTest < Test::Unit::TestCase
       end
 
       import_path = "github.com/apache/arrow/go/v#{@snapshot_major_version}"
-      hunks = []
       if release_type == :major
-        lines = File.readlines(path, chomp: true)
-        target_lines = lines.each_with_index.select do |line, i|
-          line.include?(import_path)
-        end
-        next if target_lines.empty?
-        n_context_lines = 3 # The default of Git's diff.context
-        target_hunks = [[target_lines.first[0]]]
-        previous_i = target_lines.first[1]
-        target_lines[1..-1].each do |line, i|
-          if i - previous_i < n_context_lines
-            target_hunks.last << line
-          else
-            target_hunks << [line]
-          end
-          previous_i = i
-        end
-        target_hunks.each do |lines|
-          hunk = []
-          lines.each do |line,|
-            hunk << "-#{line}"
-          end
-          lines.each do |line|
+        hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
+          if line.include?(import_path)
             new_line = line.gsub("v#{@snapshot_major_version}") do
               "v#{@next_major_version}"
             end
-            hunk << "+#{new_line}"
+            [line, new_line]
+          else
+            [nil, nil]
           end
-          hunks << hunk
         end
+      else
+        hunks = []
       end
       if path == "go/parquet/writer_properties.go"
         hunks << [
@@ -287,18 +269,17 @@ class PostBumpVersionsTest < Test::Unit::TestCase
     end
 
     Dir.glob("java/**/pom.xml") do |path|
-      version = "<version>#{@snapshot_version}</version>"
-      lines = File.readlines(path, chomp: true)
-      target_lines = lines.grep(/#{Regexp.escape(version)}/)
-      hunks = []
-      target_lines.each do |line|
-        new_line = line.gsub(@snapshot_version) do
-          @next_snapshot_version
+      hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
+        if line.include?("<version>#{@snapshot_version}</version>")
+          new_line = line.gsub(@snapshot_version) do
+            @next_snapshot_version
+          end
+          [line, new_line]
+        elsif line.include?("<project.build.outputTimestamp>")
+          [line, normalize_pom_xml_output_timestamp(line)]
+        else
+          [nil, nil]
         end
-        hunks << [
-          "-#{line}",
-          "+#{new_line}",
-        ]
       end
       expected_changes << {hunks: hunks, path: path}
     end
diff --git a/dev/release/test-helper.rb b/dev/release/test-helper.rb
index 3b2c3aa6e5..4238a534d1 100644
--- a/dev/release/test-helper.rb
+++ b/dev/release/test-helper.rb
@@ -83,15 +83,59 @@ module GitRunnable
       when /\A@@/
         in_hunk = true
         diffs.last[:hunks] << []
-      when /\A[-+]/
+      when /\A-/
         next unless in_hunk
         diffs.last[:hunks].last << line.chomp
+      when /\A\+/
+        next unless in_hunk
+        diffs.last[:hunks].last << normalize_added_line(line.chomp)
       end
     end
     diffs.sort_by do |diff|
       diff[:path]
     end
   end
+
+  def generate_hunks(lines)
+    git_diff_context = 3 # The default of Git's diff.context
+    max_lines_for_same_hunk = git_diff_context * 2 + 1
+    previous_i = nil
+    grouped_change_blocks = []
+    lines.each_with_index do |line, i|
+      deleted, added = yield(line)
+      next if deleted.nil? and added.nil?
+      if previous_i.nil? or (i - previous_i) > max_lines_for_same_hunk
+        grouped_change_blocks << []
+      end
+      if i - 1 != previous_i
+        grouped_change_blocks.last << []
+      end
+      grouped_change_blocks.last.last << [deleted, added]
+      previous_i = i
+    end
+    grouped_change_blocks.collect do |change_blocks|
+      hunk = []
+      change_blocks.each do |continuous_changes|
+        continuous_changes.each do |deleted, _|
+          hunk << "-#{deleted}" if deleted
+        end
+        continuous_changes.each do |_, added|
+          hunk << "+#{added}" if added
+        end
+      end
+      hunk
+    end
+  end
+
+  def normalize_pom_xml_output_timestamp(line)
+    line.gsub(/<project\.build\.outputTimestamp>.+?</) do
+      "<project.build.outputTimestamp>1970-01-01T00:00:00Z<"
+    end
+  end
+
+  def normalize_added_line(line)
+    normalize_pom_xml_output_timestamp(line)
+  end
 end
 
 module VersionDetectable
```

**Comment by jbonofre** [comment]
> When is `<project.build.outputTimestamp>` updated? Does `mvn versions:set` update it?
> 
> It seems that our `dev/release/01-prepare.sh` updates it: https://github.com/apache/arrow/actions/runs/7263564652/job/19789165030?pr=39215#step:7:24478

`outputTimestamp` should be always the same to guarantee the same artifact accross build. I'm doing a pass on actions to be sure we are clean here.

**Comment by jbonofre** [comment]
> It seems that updated value is always the current time: https://github.com/mojohaus/versions/pull/522/files#diff-544db8b7e4906f57cb6c1bd43ba1100760f876da22b46b2ca381ab374cb169faR530

On a specific version, the `outputTimestamp` has to be the same. So settings the `outputTimestamp` when using `verisons:set` is fine. The most important is to have a "stable" `outputTimestamp` across builds of the same version.

**Comment by jbonofre** [comment]
> How about this for "Dev /Source Release and Merge Script" failures?
> 
> ```diff
> diff --git a/dev/release/01-prepare-test.rb b/dev/release/01-prepare-test.rb
> index 8fb23f45f0..f4bf01f1f6 100644
> --- a/dev/release/01-prepare-test.rb
> +++ b/dev/release/01-prepare-test.rb
> @@ -264,18 +264,17 @@ class PrepareTest < Test::Unit::TestCase
>      end
>  
>      Dir.glob("java/**/pom.xml") do |path|
> -      version = "<version>#{@snapshot_version}</version>"
> -      lines = File.readlines(path, chomp: true)
> -      target_lines = lines.grep(/#{Regexp.escape(version)}/)
> -      hunks = []
> -      target_lines.each do |line|
> -        new_line = line.gsub(@snapshot_version) do
> -          @release_version
> +      hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
> +        if line.include?("<version>#{@snapshot_version}</version>")
> +          new_line = line.gsub(@snapshot_version) do
> +            @release_version
> +          end
> +          [line, new_line]
> +        elsif line.include?("<project.build.outputTimestamp>")
> +          [line, normalize_pom_xml_output_timestamp(line)]
> +        else
> +          [nil, nil]
>          end
> -        hunks << [
> -          "-#{line}",
> -          "+#{new_line}",
> -        ]
>        end
>        expected_changes << {hunks: hunks, path: path}
>      end
> diff --git a/dev/release/post-11-bump-versions-test.rb b/dev/release/post-11-bump-versions-test.rb
> index 4b6933d610..08a900f71c 100644
> --- a/dev/release/post-11-bump-versions-test.rb
> +++ b/dev/release/post-11-bump-versions-test.rb
> @@ -244,37 +244,19 @@ class PostBumpVersionsTest < Test::Unit::TestCase
>        end
>  
>        import_path = "github.com/apache/arrow/go/v#{@snapshot_major_version}"
> -      hunks = []
>        if release_type == :major
> -        lines = File.readlines(path, chomp: true)
> -        target_lines = lines.each_with_index.select do |line, i|
> -          line.include?(import_path)
> -        end
> -        next if target_lines.empty?
> -        n_context_lines = 3 # The default of Git's diff.context
> -        target_hunks = [[target_lines.first[0]]]
> -        previous_i = target_lines.first[1]
> -        target_lines[1..-1].each do |line, i|
> -          if i - previous_i < n_context_lines
> -            target_hunks.last << line
> -          else
> -            target_hunks << [line]
> -          end
> -          previous_i = i
> -        end
> -        target_hunks.each do |lines|
> -          hunk = []
> -          lines.each do |line,|
> -            hunk << "-#{line}"
> -          end
> -          lines.each do |line|
> +        hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
> +          if line.include?(import_path)
>              new_line = line.gsub("v#{@snapshot_major_version}") do
>                "v#{@next_major_version}"
>              end
> -            hunk << "+#{new_line}"
> +            [line, new_line]
> +          else
> +            [nil, nil]
>            end
> -          hunks << hunk
>          end
> +      else
> +        hunks = []
>        end
>        if path == "go/parquet/writer_properties.go"
>          hunks << [
> @@ -287,18 +269,17 @@ class PostBumpVersionsTest < Test::Unit::TestCase
>      end
>  
>      Dir.glob("java/**/pom.xml") do |path|
> -      version = "<version>#{@snapshot_version}</version>"
> -      lines = File.readlines(path, chomp: true)
> -      target_lines = lines.grep(/#{Regexp.escape(version)}/)
> -      hunks = []
> -      target_lines.each do |line|
> -        new_line = line.gsub(@snapshot_version) do
> -          @next_snapshot_version
> +      hunks = generate_hunks(File.readlines(path, chomp: true)) do |line|
> +        if line.include?("<version>#{@snapshot_version}</version>")
> +          new_line = line.gsub(@snapshot_version) do
> +            @next_snapshot_version
> +          end
> +          [line, new_line]
> +        elsif line.include?("<project.build.outputTimestamp>")
> +          [line, normalize_pom_xml_output_timestamp(line)]
> +        else
> +          [nil, nil]
>          end
> -        hunks << [
> -          "-#{line}",
> -          "+#{new_line}",
> -        ]
>        end
>        expected_changes << {hunks: hunks, path: path}
>      end
> diff --git a/dev/release/test-helper.rb b/dev/release/test-helper.rb
> index 3b2c3aa6e5..4238a534d1 100644
> --- a/dev/release/test-helper.rb
> +++ b/dev/release/test-helper.rb
> @@ -83,15 +83,59 @@ module GitRunnable
>        when /\A@@/
>          in_hunk = true
>          diffs.last[:hunks] << []
> -      when /\A[-+]/
> +      when /\A-/
>          next unless in_hunk
>          diffs.last[:hunks].last << line.chomp
> +      when /\A\+/
> +        next unless in_hunk
> +        diffs.last[:hunks].last << normalize_added_line(line.chomp)
>        end
>      end
>      diffs.sort_by do |diff|
>        diff[:path]
>      end
>    end
> +
> +  def generate_hunks(lines)
> +    git_diff_context = 3 # The default of Git's diff.context
> +    max_lines_for_same_hunk = git_diff_context * 2 + 1
> +    previous_i = nil
> +    grouped_change_blocks = []
> +    lines.each_with_index do |line, i|
> +      deleted, added = yield(line)
> +      next if deleted.nil? and added.nil?
> +      if previous_i.nil? or (i - previous_i) > max_lines_for_same_hunk
> +        grouped_change_blocks << []
> +      end
> +      if i - 1 != previous_i
> +        grouped_change_blocks.last << []
> +      end
> +      grouped_change_blocks.last.last << [deleted, added]
> +      previous_i = i
> +    end
> +    grouped_change_blocks.collect do |change_blocks|
> +      hunk = []
> +      change_blocks.each do |continuous_changes|
> +        continuous_changes.each do |deleted, _|
> +          hunk << "-#{deleted}" if deleted
> +        end
> +        continuous_changes.each do |_, added|
> +          hunk << "+#{added}" if added
> +        end
> +      end
> +      hunk
> +    end
> +  end
> +
> +  def normalize_pom_xml_output_timestamp(line)
> +    line.gsub(/<project\.build\.outputTimestamp>.+?</) do
> +      "<project.build.outputTimestamp>1970-01-01T00:00:00Z<"
> +    end
> +  end
> +
> +  def normalize_added_line(line)
> +    normalize_pom_xml_output_timestamp(line)
> +  end
>  end
>  
>  module VersionDetectable
> ```

Yeah, as said on a comment, I fixed the maven enforcer issue, I'm not on the merge action issue (testing locally).

**Comment by kou** [comment]
> Yeah, as said on a comment, I fixed the maven enforcer issue, I'm not on the merge action issue (testing locally).

"the maven enforcer issue" is https://github.com/apache/arrow/pull/39215#issuecomment-1862842437 , right?

I'm focusing on "Dev / Source Release and Merge Script on ubuntu-latest" CI job https://github.com/apache/arrow/actions/runs/7263564652/job/19789165030?pr=39215 not "the maven enforcer issue".

**Comment by jbonofre** [comment]
@kou Maven enforcer fix is part of this PR: https://github.com/apache/arrow/pull/39215/files#diff-d8225ebcfc11b480a6e4f54e183b67c3ead51635a167c106d928c2abf1f9ef66R31 and https://github.com/apache/arrow/pull/39215/files#diff-65ce66a6aeda30354aedbae081d0411a4ae6561f9d4745b86117ea7e3eb1839aR29

I'm on the "Release and Merge script" investigation right now 😄

**Comment by kou** [comment]
> I'm on the "Release and Merge script" investigation right now 😄

https://github.com/apache/arrow/pull/39215#issuecomment-1863659173 will fix it.

**Comment by jbonofre** [comment]
Thanks. Let me try.

File: dev/release/test-helper.rb [hunk 3]
```diff
@@ -83,15 +83,60 @@ def parse_patch(patch)
       when /\A@@/
         in_hunk = true
         diffs.last[:hunks] << []
-      when /\A[-+]/
+      when /\A-/
         next unless in_hunk
         diffs.last[:hunks].last << line.chomp
+			when /\A\+/
+				next unless in_hunk
+				diffs.last[:hunks].last << normalize_added_line(line.chomp)
```
Comment (kou) [inline dev/release/test-helper.rb:91]: I'll push indent fix for `dev/release/*.rb` changes.

Comment (jbonofre) [reply dev/release/test-helper.rb]: My bad (bad vim configuration). Let me fix the indent. Sorry about that.

Comment (kou) [reply dev/release/test-helper.rb]: No problem. I've fixed the indent.

**Review by kou — APPROVED** [review]
+1 for `dev/release/` part.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

File: dev/release/test-helper.rb (hunk 3 above)
Comment (jbonofre) [reply dev/release/test-helper.rb]: I pushed the fix as well.

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by lidavidm** [comment]
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.6.0:single (default-cli) on project arrow-bom: Error reading assemblies: No assembly descriptors found. -> [Help 1]
```

hmm

**Comment by jbonofre** [comment]
Weird. Let me check.

**Comment by jbonofre** [comment]
@lidavidm the assembly plugin failed on the BoM. I'm checking.

**Comment by raulcd** [comment]
@jbonofre the crossbow task uses * to match string patterns. You can use `*java*` and it will work

**Comment by jbonofre** [comment]
@raulcd thanks, my bad (typo mistake 😄 )

**Comment by jbonofre** [comment]
@github-actions crossbow submit *java*

File: dev/release/test-helper.rb (hunk 3 above)
Comment (kou) [reply dev/release/test-helper.rb]: Oh, why...?
You broke it again...: https://github.com/apache/arrow/compare/6d91fddd6ce3bfa28cbb25421cb671b1b737147e..cb22662007ecada122cd6d441546e3385c31e906#diff-7655978b5793694e21d21f4c483847ed0004cd1035ebd91a87c862c246bbb83a

You don't need to squash all your commits in a PR because we use "squash & merge".

File: dev/release/test-helper.rb [hunk 4]
```diff
@@ -83,15 +83,60 @@ def parse_patch(patch)
...
+  def normalize_pom_xml_output_timestamp(line)
+    line.gsub(/<project\.build\.outputTimestamp>.+?</) do
+      "<project.build.outputTimestamp>2023-12-13T00:00:00Z<"
+    end
+  end
+
+  def normalize_added_line(line)
+    normalize_pom_xml_output_timestamp(line)
+  end
+
```
Comment (kou) [inline dev/release/test-helper.rb:139]:
```suggestion
```

File: dev/release/test-helper.rb [hunk 5]
```diff
@@ -83,15 +83,60 @@ def parse_patch(patch)
       when /\A@@/
         in_hunk = true
         diffs.last[:hunks] << []
-      when /\A[-+]/
+      when /\A-/
         next unless in_hunk
         diffs.last[:hunks].last << line.chomp
+		  when /\A\+/
+		    next unless in_hunk
+			diffs.last[:hunks].last << normalize_added_line(line.chomp)
```
Comment (kou) [inline dev/release/test-helper.rb:91]:
```suggestion
      when /\A\+/
        next unless in_hunk
        diffs.last[:hunks].last << normalize_added_line(line.chomp)
```

File: dev/release/test-helper.rb [hunk 6]
```diff
@@ -83,15 +83,60 @@ def parse_patch(patch)
...
+          hunk << "+#{added}" if added
+        end
+      end
+      hunk
+    end
+  end
+
+  def normalize_pom_xml_output_timestamp(line)
+    line.gsub(/<project\.build\.outputTimestamp>.+?</) do
+      "<project.build.outputTimestamp>2023-12-13T00:00:00Z<"
```
Comment (kou) [inline dev/release/test-helper.rb:132]: Why did you change the normalized timestamp to `2023-12-13` from `1970-01-01`?
I think that special date (`1970-01-01` is a special date because UNIX time uses it as epoch) is better for easy to understand that we normalize our timestamps.

**Comment by jbonofre** [comment]
I will complete the pr tomorrow (finishing Apache Karaf release right now ;)).

File: dev/release/test-helper.rb (hunk 3 above)
Comment (jbonofre) [reply dev/release/test-helper.rb]: Ah ok. On most of other Apache projects I'm working we do a single commit to use simple merge. Sorry. I will fix indent.

File: dev/release/test-helper.rb (hunk 6 above)
Comment (jbonofre) [reply dev/release/test-helper.rb:132]: The most important is to have the same output timestamp. I can use 1T 1970. I will update it.

Comment (jbonofre) [reply dev/release/test-helper.rb]: I updated to use `1980-01-01T00:00:02Z` to avoid issue with `maven-assembly-plugin`.

**Comment by jbonofre** [comment]
@github-actions crossbow submit *java*

File: dev/release/test-helper.rb (hunk 6 above)
Comment (kou) [reply dev/release/test-helper.rb]: We don't need to use the special timestamp in `pom.xml`. (Timestamps in `pom.xml` are updated automatically in our release process, right? https://github.com/apache/arrow/pull/39215#issuecomment-1863620913 )
We normalize timestamps in `pom.xml` for testing by this code.

Comment (jbonofre) [reply dev/release/test-helper.rb]: I think it's better (for both visibility and testing) to have `project.build.outputTimestamp` in the `pom.xml`. Or at least, I can replace it with a comment documenting that `project.build.outputTimestamp` is populating by the `dev/release/test-helper.rb`. Thoughts ?

**Comment by jbonofre** [comment]
OK, I understand what happens with `ci/scripts/java_full_build.sh` script: it explicitly call `assembly:single` goal. I'm fixing that :)

File: dev/release/test-helper.rb (hunk 6 above)
Comment (kou) [reply dev/release/test-helper.rb]: Hmm, it seems that we have miscommunication. Here are what I think:
* We need the `project.build.outputTimestamp` element in the `pom.xml`.
* But we don't need to use a special timestamp for the element in the `pom.xml` like you did in the first commit in this PR.

`dev/release/test-helper.rb` updates the element's timestamp because it runs partial our release process. And the process updates `project.build.outputTimestamp`'s timestamp.

Comment (jbonofre) [reply dev/release/test-helper.rb]: To summarise:
1. We need `project.build.outputTimestamp` in the `pom.xml` as it's used by maven plugins to ensure the build is reproducible.
2. `project.build.outputTimestamp` value has to be fixed (not changing at each build). The value doesn't matter if it's always the same. That's the purpose of the setting the value (it's what I did in all other Apache projects where I'm working on).
3. We can't use `1970-01-01` else maven-assembly-plugin complains, that's why I used the first allowed value in the range which is `1980-01-01T00:00:02Z`. So I used the same in the `test-helper.rb` to be in sync.

Is it OK for you ?

Thanks

Comment (kou) [reply dev/release/test-helper.rb]: >     1. We need `project.build.outputTimestamp` in the `pom.xml` as it's used by maven plugins to ensure the build is reproducible.

Yes. I think so too.

>     2. `project.build.outputTimestamp` value has to be fixed (not changing at each build). The value doesn't matter if it's always the same. That's the purpose of the setting the value (it's what I did in all other Apache projects where I'm working on).

Yes. I think so too.
 
>     3. We can't use `1970-01-01` else maven-assembly-plugin complains, that's why I used the first allowed value in the range which is `1980-01-01T00:00:02Z`. So I used the same in the `test-helper.rb` to be in sync.

Yes and no. We don't need to use special timestamps such as `1970-...` and `1980-...` in `pom.xml`. `test-helper.rb` normalizes all timestamps such as  `2023-12-13T00:00:00Z` in `pom.xml`. So we don't need to synchronize timestamps in `pom.xml` and `test-helper.rb`. So we can use  `2023-12-13T00:00:00Z` in `pom.xml` like you did.

**Comment by jbonofre** [comment]
@github-actions crossbow submit *java*

File: java/adapter/avro/pom.xml [hunk 7]
```diff
@@ -25,6 +25,10 @@
   <description>(Contrib/Experimental) A library for converting Avro data to Arrow data.</description>
   <url>http://maven.apache.org</url>
 
+  <properties>
+    <project.build.outputTimestamp>1980-01-01T00:00:02Z</project.build.outputTimestamp>
```
Comment (pitrou) [inline java/adapter/avro/pom.xml:29]: Does this force the output timestamp for all builds regardless of build options?

Comment (jbonofre) [reply java/adapter/avro/pom.xml:29]: Hey @pitrou . The `project.build.outputTimestamp` property is used by several Maven plugins to ensure the consistency of the generated artifacts (assembly, jar, source, ...). They "hook" in their goals on phases where it's required. So, yes, for instance, using `package` phase will set the output timestamp on the artifacts. Does it answer your question (not sure I understood your question correctly ;) ) ?

Comment (pitrou) [reply java/adapter/avro/pom.xml:29]: My question is, is this only enabled for specific build variants or options? Or would I get those fixed timestamps if I do a local Java build?

Comment (jbonofre) [reply java/adapter/avro/pom.xml:29]: It applies to any build including your local one. So you can have a reproducible build locally.

Comment (pitrou) [reply java/adapter/avro/pom.xml:29]: Wouldn't it be confusing in some situations then, to have files dated to some random past value?

Comment (jbonofre) [reply java/adapter/avro/pom.xml:29]: It's only the archives/artifacts. And it's good practice generally speaking (we do that in most of Apache projects).

Comment (pitrou) [reply java/adapter/avro/pom.xml:29]: Fair enough. I was expecting this to be enabled selectively when building public packages.

**Comment by jbonofre** [comment]
@github-actions crossbow submit *java*

**Comment by jbonofre** [comment]
@lidavidm I'm investigating why GH Actions are not happy now (weird). Sorry for the delay.

**Comment by jbonofre** [comment]
@lidavidm do you have an idea why `C++ / AMD64 Conda C++ AVX2`, `Java JNI / AMD64 manylinux2014 Kava JNI`, etc jobs are failing ?
It looks weird to me as it seems related to docker container execution.

File: cpp/submodules/parquet-testing
Comment (lidavidm) [inline cpp/submodules/parquet-testing:1]: Can you revert this change? That might be why.

Comment (jbonofre) [reply cpp/submodules/parquet-testing:1]: Ah yeah, possible indeed. Let me try thanks.

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by lidavidm** [comment]
Hmm well Archery doesn't like me but it seems CI still has a few issues even after a re-run.

**Comment by jbonofre** [comment]
@lidavidm yup, to be honest, I don't know why it's failing 😄 I will probably try on `main` branch to compare.
I will do a new pass again, maybe I missed something 😄

**Comment by kou** [comment]
@assignUser Could you take a look at https://github.com/apache/arrow/pull/39215#issuecomment-1900481478 and https://github.com/apache/arrow/pull/39215#issuecomment-1900482331 ?

#39610 is related.

**Comment by assignUser** [comment]
@kou hm weird asf membership is public for David so the bot should pick it up... if this keeps happening we may need to add a non gh api based method to identify committers (e.g. committer list, IIRC we have a yaml one around somehwere right?)

**Comment by lidavidm** [comment]
I fixed my membership after that. You may want to update the bot's error message to tell people what to do, I only knew because I happened to see someone else ask the same question in a private channel.

**Comment by jbonofre** [comment]
I'm rebasing and doing new tests on this one.

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by jbonofre** [comment]
@lidavidm as I'm back on Arrow, I'm resuming several PRs/works I have in progress 😄 Thanks !

**Comment by vibhatha** [comment]
@jbonofre need any help here?

**Comment by jbonofre** [comment]
I'm back on Arrow. I will rebase and retest this PR.

**Comment by kou** [comment]
We're moving `java/` to https://github.com/apache/arrow-java .
Could you open a new PR in apache/arrow-java?
